iml-gui/crate/src/components/command_modal.rs

```diff
@@ -284,3 +537,601 @@ fn close_button() -> Node<Msg> {
+        assert_eq!(is_subset(&vec![5, 1], &all), false);
+    }
+
+    #[test]
+    fn test_interpret_click() {
+        // test transitions, we simulate the clicks on various elements
+
+        let pair = (Select::Command(77), false);
+
+        let pair = interpret_click(&pair.0, TypedId::Job(162));
+        assert_eq!(&pair, &(Select::CommandJob(77, vec![162]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Job(163));
+        assert_eq!(&pair, &(Select::CommandJob(77, vec![162, 163]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Step(309));
+        assert_eq!(&pair, &(Select::CommandJobSteps(77, vec![162, 163], vec![309]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Step(308));
+        assert_eq!(
+            &pair,
+            &(Select::CommandJobSteps(77, vec![162, 163], vec![308, 309]), true)
+        );
+
+        let pair = interpret_click(&pair.0, TypedId::Job(163));
+        assert_eq!(&pair, &(Select::CommandJobSteps(77, vec![162], vec![308, 309]), false));
+
+        let pair = interpret_click(&pair.0, TypedId::Job(162));
+        assert_eq!(&pair, &(Select::Command(77), false));
+    }
+
+    /// There could the the problem, when `schedule_fetch_tree` made the api requests
+    /// in one order, but during delays or something the responses may come in different
+    /// order. In all such cases, however, model should remain consistent,
+    /// and the test checks this.
+    #[test]
+    fn test_async_handlers_consistency() {
+        let mut rng = Xoroshiro64Star::seed_from_u64(485369);
+        let db = build_db();
+
+        let mut model = Model::default();
+        let cmd_ids = db.all_cmds.iter().map(|x| x.id).collect::<Vec<_>>();
+        let mut permutation = vec![1, 2, 3];
+        let selects = vec![
+            Select::None,
+            Select::Command(2),
+            Select::Command(1),
+            Select::Command(3),
+            Select::Command(4),
+            Select::CommandJob(1, vec![10]),
+            Select::CommandJob(1, vec![10, 11]),
+            Select::CommandJobSteps(1, vec![10, 11], vec![21]),
+            Select::CommandJobSteps(1, vec![10, 11], vec![20, 21]),
+        ];
+        for select in selects {
+            let cmd_id = cmd_ids[rng.next_u32() as usize % cmd_ids.len()];
+            let (c, j, s) = prepare_subset(&db, cmd_id);
+            model.clear();
+            model.select = select.clone();
+            model.assign_commands(&c);
+            model.assign_jobs(&j);
+            model.assign_steps(&s);
+            let expected_cmd = model.commands_view.clone();
+            let expected_jobs = model.jobs_graph.clone();
+            let expected_steps = model.steps_view.clone();
+
+            model.clear();
+            model.select = select.clone();
+            shuffle(&mut permutation, &mut rng);
+
+            // we simulate, that FetchCommands, FetchJobs and FetchSteps come in arbitrary order
```

Comment:
This test must be unstable, if the problem really existed. I can't see how assign_foo may depend on the order they are invoked.

There are only 3! = 6 permutations, and one of them is already covered above, so you can just write them all down here without RNG. But still I don't see the problem this test is trying catch.

Reply:
If you print out the permutation on each pass, you will see this every run (i. e. it is not random)

```
[2, 3, 1]
[3, 1, 2]
[2, 3, 1]
[3, 2, 1]
[3, 2, 1]
[3, 2, 1]
[1, 3, 2]
[3, 1, 2]
[3, 2, 1]
```

And you can see that [2,1,3] is missed :D

Reply:
This means, just that the number of `select`-s is too low, given enough generated test cases the proof of the correctness could be made strong enough, i.e. for every `ε` the probability to bypass the false case could be made smaller than `ε`. Agree, this needs a small improvement.